apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransactionOptions.cs

```diff
@@ -100,5 +101,41 @@ internal int EffectiveCommitTimeout(SpannerConnection spannerConnection) =>
         TransactionMode.ReadOnly => Priority.Unspecified,
         _ => CommitPriority ?? Priority.Unspecified
     };
-        
+
+    /// <summary>
+    /// The transaction tag to use for <see cref="SpannerTransaction"/>
+    /// This value will be ignored by read-only transactions.
+    /// Changes to this value will be ignored by the transaction after it has executed statements.
+    /// </summary>
+    public string Tag { get; set; }
+    private string _beforeFirstStatementTag;
+
+    internal string EffectiveTag(TransactionMode transactionMode, bool hasExecutedStatements)
+    {
+        MaybeAfterStatementExecution(hasExecutedStatements, transactionMode);
+        return hasExecutedStatements ? _beforeFirstStatementTag : EffectiveTag(transactionMode);
+    }
+
+    private string EffectiveTag(TransactionMode transactionMode) => transactionMode switch
+    {
+        TransactionMode.ReadOnly => "",
+        _ => Tag ?? ""
+    };
+
+    private bool _hasExecutedStatements = false;
+    private readonly object _statementExecutionLock = new object();
+    private void MaybeAfterStatementExecution(bool hasExecutedStatements, TransactionMode transactionMode)
+    {
+        lock (_statementExecutionLock)
+        {
+            // This should never fail.
```

Comment:
Because this should be called exactly once with hasExecutedStatements being true, and that should be the first call? (If so, maybe put that in the comment.)